scalyr_agent/log_processing.py

```diff
@@ -1447,6 +1453,36 @@ def __init__(self, file_path, config, log_config, close_when_staleness_exceeds=N
 
         self.__disable_processing_new_bytes = config.disable_processing_new_bytes
 
+    def finish( self ):
+        """
+        Marks the log processor as finished.
+
+        When a log processor is `finished`, then it can be scheduled for removal from the
+        main copy manager once it no longer has any bytes available for reading.
+        """
+        self.__lock.acquire()
+        try:
+            self.__is_finished = True
+        finally:
+            self.__lock.release()
+
+    def is_finished( self ):
+        """
+        Returns whether or not this processor is in the finished state, *and* there are no bytes
+        available for reading.
+
+        If this method returns True, then the log processor can be removed from the copy manager
+        """
+        # We only lock for __is_finished.  The other fields are only modified on the main thread
+        self.__lock.acquire()
+        try:
+            # __is_finished means the log has been marked as finished
+            # __prev_bytes_read == 0 means the last call to perform_processing didn't have any bytes for this processor
+            # __total_bytes_pending == 0 means there are no more bytes available from the iterator
+            return self.__is_finished and self.__prev_bytes_read == 0 and self.__log_file_iterator.available == 0
```

Comment:
Note, I believe there is a race condition here.  It could be that there is a request in-flight that has the last bytes from the file in it... The `available` would still return `0` even though the request is in flight.

Now, this method will only be called (presumably) by the `CopyManager` so it could be that the `CopyManager` only invokes it when it knows there are no requests in flight... but that means we'd have to document this method as such and make sure we don't violate it.  With the changes I propose below, I think we can avoid having to do all of that.